src/esp/gfx/Renderer.cpp

```diff
@@ -27,16 +34,236 @@ namespace Mn = Magnum;
 namespace esp {
 namespace gfx {
 
+struct BackgroundRenderThread {
+  BackgroundRenderThread(WindowlessContext* context)
+      : context_{context}, done_{0}, sgLock_{0}, barrierVal_{0} {
+    context_->release();
+    t = std::thread(&BackgroundRenderThread::run, this);
+
+    jobsWaiting_ = 1;
+    threadStarted_ = true;
+    waitThread();
+    context_->makeCurrent();
+  }
+
+  void startThread() {
+    barrierVal_.fetch_xor(1, std::memory_order_relaxed);
+    std::atomic_thread_fence(std::memory_order_release);
+    cv_.notify_all();
+    threadStarted_ = true;
+  }
+
+  ~BackgroundRenderThread() {
+    task_ = Task::Exit;
+    startThread();
+    t.join();
+  }
+
+  static void spinLock(const std::atomic<int>& lk, int val) {
+    while (lk.load(std::memory_order_acquire) != val)
+      asm volatile("pause" ::: "memory");
+  }
+
+  void waitThread() {
+    CORRADE_INTERNAL_ASSERT(threadStarted_ || jobsWaiting_ == 0);
+    spinLock(done_, jobsWaiting_);
+
+    done_.store(0);
+    jobsWaiting_ = 0;
+    threadStarted_ = false;
+  }
+
+  void waitSG() { spinLock(sgLock_, 0); }
+
+  void submitJob(sensor::VisualSensor& sensor,
+                 scene::SceneGraph& sceneGraph,
+                 const Mn::MutableImageView2D& view,
+                 RenderCamera::Flags flags) {
+    jobs_.emplace_back(std::ref(sensor), std::ref(sceneGraph), std::cref(view),
+                       flags);
+    jobsWaiting_ += 1;
+  }
+
+  void startJobs() {
+    task_ = Task::Render;
+    sgLock_.store(1, std::memory_order_relaxed);
+    startThread();
+  }
+
+  void releaseContext() {
+    task_ = Task::ReleaseContext;
+
+    startThread();
+
+    jobsWaiting_ = 1;
+    waitThread();
+  }
+
+  int threadRender() {
+    if (!threadOwnsContext_) {
+      VLOG(1)
+          << "BackgroundRenderThread:: Background thread acquired GL Context";
+      context_->makeCurrentPlatform();
+      Mn::GL::Context::makeCurrent(threadContext_);
+      threadOwnsContext_ = true;
+    }
+
+    std::vector<std::vector<RenderCamera::DrawableTransforms>> jobTransforms(
+        jobs_.size());
+
+    for (int i = 0; i < jobs_.size(); ++i) {
+      auto& job = jobs_[i];
+      sensor::VisualSensor& sensor = std::get<0>(job);
+      scene::SceneGraph& sg = std::get<1>(job);
+      RenderCamera::Flags flags = std::get<3>(job);
+
+      auto* camera = sensor.getRenderCamera();
+      jobTransforms[i].reserve(sg.getDrawableGroups().size());
+      for (auto& it : sg.getDrawableGroups()) {
+        it.second.prepareForDraw(*camera);
+        auto transforms = camera->drawableTransformations(it.second);
```

Comment:
So you are reading sceneGraph node transforms on this background thread. Isn't Bullet potentially writing those transforms on the main thread? This writing happens via MotionState.cpp MotionState::setWorldTransform, which Bullet calls internally because we provide this motionState when we construct a Bullet rigid body in BulletRigidObject.

Reply:
I probably should have put that change in this PR, but I stuck that in this one: https://github.com/facebookresearch/habitat-sim/pull/1100 (made testing easier).  That added a defer feature to MotionState so that Bullet won't update the SceneGraph until the call to `updateNodes` which is why where is ` renderer_->waitSG()` before that